src/buffer/buffer.rs

```diff
@@ -134,17 +153,31 @@ impl Buffer {
     /// ```
     #[track_caller]
     pub fn index_of(&self, x: u16, y: u16) -> usize {
-        debug_assert!(
-            x >= self.area.left()
-                && x < self.area.right()
-                && y >= self.area.top()
-                && y < self.area.bottom(),
-            "Trying to access position outside the buffer: x={x}, y={y}, area={:?}",
-            self.area
-        );
-        ((y - self.area.y) * self.area.width + (x - self.area.x)) as usize
+        self.try_index_of(x, y)
+            .unwrap_or_else(|| Self::panic_on_wrong_position(x, y, self.area))
+    }
+    fn panic_on_wrong_position(x: u16, y: u16, area: Rect) -> ! {
+        panic!("Trying to access position outside the buffer: x={x}, y={y}, area={area:?}",)
+    }
+    /// Returns an option containing the index in the `Vec<Cell>` for the given global (x, y)
+    /// coordinates, or `None` if the position is outside the buffer.
+    /// Global coordinates are offset by the Buffer's area offset (`x`/`y`).
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use ratatui::prelude::*;
+    /// let rect = Rect::new(200, 100, 10, 10);
+    /// let buffer = Buffer::empty(rect);
+    /// // Global coordinates to the top corner of this buffer's area
+    /// assert_eq!(buffer.try_index_of(200, 100), Some(0));
+    /// ```
+    #[must_use]
+    pub fn try_index_of(&self, x: u16, y: u16) -> Option<usize> {
```

Comment:
I'd prefer not to add more cruft. The fact that the underlying storage of the cells is a Vec is an implementation detail, not something that should generally be relied on by callers. `index_of` should be pub(crate) as should anything in buffer that deals directly with the idea of the index. That's a bigger change than this however. But, adding another public method makes that more annoying.